### PR TITLE
Role category Hierarchies

### PR DESCRIPTION
**Pull request by Ryanseanlee**
Hoist P/R Checklist
-------------------

**Pull request authors:** Review and check off the below. Items that do not apply can also be
checked off to indicate they have been considered. If unclear if a step is relevant, please leave
unchecked and note in comments.

- [x] Caught up with `develop` branch as of last change.
- [ ] Added CHANGELOG entry, or determined not required.
- [ ] Reviewed for breaking changes, added `breaking-change` label + CHANGELOG if so.
- [ ] Updated doc comments / prop-types, or determined not required.
- [ ] Reviewed and tested on Mobile, or determined not required.
- [ ] Created Toolbox branch / PR, or determined not required.

**If your change is still a WIP**, please use the "Create draft pull request" option in the split
button below to indicate it is not ready yet for a final review.

> **Pull request reviewers:** when merging this P/R, please consider using a *squash commit* to
> collapse multiple intermediate commits into a single commit representing the overall feature
> change. This helps keep the commit log clean and easy to scan across releases. PRs containing a
> single commit should be *rebased* when possible.

### CODE REVIEW

File: admin/tabs/general/roles/RoleModel.ts [hunk 1]
```diff
@@ -159,8 +172,9 @@ export class RoleModel extends HoistModel {
     private groupByAction(): RecordActionSpec {
         return {
             text: 'Group By Category',
-            displayFn: ({gridModel}) => ({
-                icon: isEmpty(gridModel.groupBy) ? Icon.circle() : Icon.checkCircle()
+            displayFn: ({gridModel, record}) => ({
+                icon: isEmpty(gridModel.groupBy) ? Icon.circle() : Icon.checkCircle(),
+                disabled: this.readonly || record?.data.isGroupRow
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:177]: This seems like a reasonable action to always allow - why disable it?

File: admin/tabs/general/roles/RoleModel.ts [hunk 2]
```diff
@@ -207,6 +221,32 @@ export class RoleModel extends HoistModel {
         });
     }
 
+    private processRolesForTreeGrid(
+        roles: Omit<HoistRole, 'users' | 'directoryGroups' | 'roles'>[]
+    ) {
+        const topList = [];
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:227]: I think this method would benefit from some more specific variable names, particularly `list` and `topList`, which feel pretty ambiguous to me.  Maybe `ret` or `root` instead of `topList` and `children` instead of `list`?

File: admin/tabs/general/roles/RoleModel.ts [hunk 3]
```diff
@@ -207,6 +221,32 @@ export class RoleModel extends HoistModel {
...
+                let currCat = list.find(it => it.name === category);
+                if (!currCat) {
+                    currCat = {name: category, children: [], isGroupRow: true};
+                    currCat.id = `${id}-${currCat.name}`;
+                    id = currCat.id;
+                    list.push(currCat);
+                }
+                list = currCat.children;
+            });
+            if (list) {
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:243]: What circumstance would cause this condition to be false?

File: admin/tabs/general/roles/RoleModel.ts [hunk 4]
```diff
@@ -207,6 +221,32 @@ export class RoleModel extends HoistModel {
...
+        roles: Omit<HoistRole, 'users' | 'directoryGroups' | 'roles'>[]
+    ) {
+        const topList = [];
+        roles.forEach(role => {
+            const categories = role.category ? role.category.split('\\') : ['Uncategorized'];
+
+            let list = topList,
+                id = '';
+            categories.forEach(category => {
+                let currCat = list.find(it => it.name === category);
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:234]: Should you also be checking that `isGroupRow === true`? What if you had a role with the same name as a category?

File: admin/tabs/general/roles/RoleModel.ts (hunk 1 above)
Comment (ghsolomon) [reply admin/tabs/general/roles/RoleModel.ts:177]: Actually, does this work anymore?

File: admin/tabs/general/roles/RoleModel.ts [hunk 5]
```diff
@@ -222,25 +262,29 @@ export class RoleModel extends HoistModel {
 
     private createGridModel(): GridModel {
         return new GridModel({
+            treeMode: true,
+            treeStyle: TreeStyle.HIGHLIGHTS_AND_BORDERS,
             autosizeOptions: {mode: 'managed'},
             emptyText: 'No roles found.',
             colChooserModel: true,
-            sortBy: 'name|asc',
+            sortBy: ['isGroupRow', 'name'],
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:270]: Is it necessary to sort on `isGroupRow`? That seems confusing to me

File: admin/tabs/general/roles/RoleModel.ts [hunk 6]
```diff
@@ -279,11 +324,12 @@ export class RoleModel extends HoistModel {
...
                     omit: this.readonly
                 },
-                {field: {name: 'name', type: 'string'}},
+                {field: {name: 'name', type: 'string'}, isTreeColumn: true},
                 {field: {name: 'category', type: 'string'}, hidden: true},
                 {field: {name: 'lastUpdated', type: 'date'}, ...Col.dateTime, hidden: true},
                 {field: {name: 'lastUpdatedBy', type: 'string'}, hidden: true},
-                {field: {name: 'notes', type: 'string'}, filterable: false, flex: 1}
+                {field: {name: 'notes', type: 'string'}, filterable: false, flex: 1},
+                {field: {name: 'isGroupRow', type: 'bool'}, hidden: true, excludeFromChooser: true}
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RoleModel.ts:332]: Does this need to be a hidden column? Seems like it could just be a `Store` `Field`?

File: admin/tabs/general/roles/RolePanel.ts [hunk 7]
```diff
@@ -42,7 +42,18 @@ export const rolePanel = hoistCmp.factory({
                     }),
                     filterChooser({flex: 1})
                 ],
-                item: hframe(vframe(grid(), roleGraph()), detailsPanel())
+                item: hframe(
+                    vframe(
+                        grid({
+                            agOptions: {
```
Comment (ghsolomon) [inline admin/tabs/general/roles/RolePanel.ts:48]: I don't think this is necessary anymore - perhaps left over from old implementation?

File: cmp/grid/Grid.scss [hunk 8]
```diff
@@ -262,6 +262,11 @@
   }
 }
 
+// TODO find better place to put this
+.xh-grid-clear-background-color:not(.ag-row-selected) {
```
Comment (ghsolomon) [inline cmp/grid/Grid.scss:266]: Just curious - why was this needed?

**Comment by lbwexler** [comment]
Hi @Ryanseanlee -- could you catch this up, and indicate if all of Greg's comments have been handled?  Would love to merge this.

**Comment by Ryanseanlee** [comment]
> Hi @Ryanseanlee -- could you catch this up, and indicate if all of Greg's comments have been handled? Would love to merge this.

Hey yes definitely

File: admin/tabs/userData/roles/RoleModel.ts [hunk 9]
```diff
@@ -159,17 +173,18 @@ export class RoleModel extends HoistModel {
     private groupByAction(): RecordActionSpec {
         return {
             text: 'Group By Category',
-            displayFn: ({gridModel}) => ({
-                icon: isEmpty(gridModel.groupBy) ? Icon.circle() : Icon.checkCircle()
+            displayFn: () => ({
+                icon: this.groupingDisabled ? Icon.circle() : Icon.checkCircle(),
+                disabled: this.readonly
```
Comment (lbwexler) [inline admin/tabs/userData/roles/RoleModel.ts:178]: why can't I group/ungroup in read only mode?

File: admin/tabs/userData/roles/RoleModel.ts [hunk 10]
```diff
@@ -38,12 +38,16 @@ export class RoleModel extends HoistModel {
     @observable.ref allRoles: HoistRole[] = [];
     @observable.ref moduleConfig: RoleModuleConfig;
 
+    @bindable groupingDisabled = false;
```
Comment (lbwexler) [inline admin/tabs/userData/roles/RoleModel.ts:41]: its not that grouping is 'disabled'  its that its not actually being grouped at the moment.  Would call it -- `showInGroups' or something like that

File: admin/tabs/userData/roles/RoleModel.ts [hunk 11]
```diff
@@ -79,8 +83,9 @@ export class RoleModel extends HoistModel {
 
     @action
     setRoles(roles: HoistRole[]) {
+        const gridData = this.groupingDisabled ? roles : this.processRolesForTreeGrid(roles);
         this.allRoles = roles;
-        this.gridModel.loadData(roles);
+        this.gridModel.loadData(gridData);
```
Comment (lbwexler) [inline admin/tabs/userData/roles/RoleModel.ts:88]: I would  add an additional private method of the form l`displayRoles()`, that does the pre-processing and then loads the data in the grid.    `allRoles` should just be set once on this public setter.

File: admin/tabs/userData/roles/RoleModel.ts [hunk 12]
```diff
@@ -212,6 +227,30 @@ export class RoleModel extends HoistModel {
         });
     }
 
+    private processRolesForTreeGrid(
+        roles: Omit<HoistRole, 'users' | 'directoryGroups' | 'roles'>[]
```
Comment (lbwexler) [inline admin/tabs/userData/roles/RoleModel.ts:231]: I need someone to explain to me how this works :)

**Comment by lbwexler** [comment]
@Ryanseanlee -- please let me know when you are ready to go over?  Hopefully monday

Would be helpful if you could resolve the comments above that you have dealt with -- thanks

File: admin/tabs/general/roles/RoleModel.ts (hunk 5 above)
Comment (Ryanseanlee) [reply admin/tabs/general/roles/RoleModel.ts]: We want the role to show on top of groups if they are on the same "level."
